Feature/async choices 2

davecoates opened the pull request.

davecoates committed:
feature: testbed user ids filter

davecoates committed:
feature: isDeepEqual, add missing exports

davecoates committed:
feature: useAsyncValue on(Success|Error)

davecoates committed:
fix: cache support id of 0

davecoates committed:
fix: UrlPattern.resolve indices=false

davecoates committed:
fix: antd upgrade, fix bad refs

davecoates committed:
feature: Labeled item interface

davecoates committed:
fix: update Identifiable docs

davecoates committed:
feature: AsyncChoicesInterface and field integration

davecoates committed:
feature: useAsyncChoices

davecoates committed:
feature: SelectAsyncChoicesWidget

davecoates committed:
fix: AsyncChoice type fixes, widget fixes

davecoates committed:
test: useAsyncChoices, AsyncChoices tests

davecoates committed:
fix: remove unnecessary ref

davecoates committed:
feature: add async choices to testbed

davecoates committed:
fix: upgrade antd

davecoates committed:
fix: identifiable _pk takes precedence

davecoates committed:
fix: remove log

davecoates committed:
feature: ViewModelEndpoint

davecoates committed:
fix: ignore empty string values

davecoates committed:
feature: major doc updates

levic commented on .isort.cfg:
```diff
@@ -17,3 +17,4 @@ known_standard_library=
 known_third_party=
 	djrad
 	djrad_rest
+	allianceutils
```
This doesn't seem to be referenced anywhere in the diff -- was it a bug fix or an accidental commit?

davecoates replied:
was causing CI to fail - has since been fixed in master

levic commented on django-testbed/django-root/xenopus_frog/views.py:
```diff
@@ -30,18 +32,38 @@ class Meta:
     first_name = filters.CharFilter(lookup_expr="icontains")
     last_name = filters.CharFilter(lookup_expr="icontains")
+    keywords = filters.CharFilter(method="filter_keywords")
+    ids = ModelIdFilter(queryset=User.objects.all())
```
For consistency that should be the first field listed

levic commented on js-packages/@prestojs/ui-antd/src/getWidgetForField.ts:
```diff
@@ -46,19 +46,20 @@ export default function getWidgetForField<FieldValue, T extends HTMLElement>(
+    const widget: FieldWidgetType<any, any> | null | undefined =
+        field.choices || field.asyncChoices
+            ? choicesMapping.get(field.constructor.name) || mapping.get(field.constructor.name)
+            : mapping.get(field.constructor.name);
```
This is `(field.choices || field.asyncChoices) ? ... : `? -- it might be worth putting brackets around that to make it a bit clearer what the order of precedence is (or maybe it's just me not seeing enough of these JS patterns)

levic commented on js-packages/@prestojs/ui-antd/src/getWidgetForField.ts:
```diff
@@ -46,19 +46,20 @@ export default function getWidgetForField<FieldValue, T extends HTMLElement>(
-        ? choicesMapping.get(field.constructor.name) || mapping.get(field.constructor.name)
-        : mapping.get(field.constructor.name);
+    const widget: FieldWidgetType<any, any> | null | undefined =
+        field.choices || field.asyncChoices
```
This is `(field.choices || field.asyncChoices) ? ... : ... `? -- it might be worth putting brackets on line 50 to make it a bit clearer what the order of precedence is (or maybe it's just me not seeing enough of these JS patterns). Assuming prettification doesn't remove them.

davecoates replied:
no can do, prettier removes them

davecoates replied:
https://github.com/prettier/prettier/issues/3805

levic commented on js-packages/@prestojs/ui-antd/src/getWidgetForField.ts:
```diff
@@ -72,9 +73,10 @@ export default function getWidgetForField<FieldValue, T extends HTMLElement>(
-            ? choicesMapping.get(f.name) || mapping.get(f.name)
-            : mapping.get(f.name);
+        const widgetF: FieldWidgetType<any, any> | null | undefined =
+            field.choices || field.asyncChoices
```
As above re parentheses

levic commented on js-packages/@prestojs/util/src/comparison.ts:
```diff
@@ -17,7 +18,7 @@ function equalityCustomizer(a: any, b: any): boolean | undefined {
         return b.isEqual(a);
     }
-    // undefined return means shallowequal does normal shallow comparison
+    // undefined return means shallowequal or isEqualWith does normal shallow/deep comparison
```
Grammar there is off

levic commented on django-testbed/frontend/src/models/User.js:
```diff
@@ -1,19 +1,27 @@
+        throw new Error(
+            'Response is paginated but paginator was not defined. Do you need to use PaginatedEndpoint instead?'
+        );
+    }
```
For a normal app you'd presumably either know whether or not pagination was going to be present on the backend.

Is this something that should perhaps be rolled into presto core and printed as a console warning in dev?

davecoates replied:
This is obsoleted by #68 which removes this check (because it's dependant on how pagination is implemented)

I couldn't think how to do the check generically so left it out but may be worth revisiting. Possible a catch all on errors that says something like 'There was a problem transforming response data. Is the response paginated?' if it fails on an unpaginated Endpoint

levic commented on django-testbed/frontend/src/models/User.js:
```diff
@@ -24,7 +32,102 @@ function transformAndCacheUser(data) {
+        transformResponseBody: transformAndCacheUser,
+        method: 'post',
+    }),
+};
```
The `endpoints` definition above we'd put in the code generator?

davecoates replied:
If not initially then long term I think yes

levic commented on js-packages/@prestojs/ui-antd/src/widgets/SelectAsyncChoiceWidget.tsx:
```diff
@@ -0,0 +1,468 @@
+};
+
+/**
+ * @expand-properties Most [Select](https://next.ant.design/components/select/) props in addition to the below. Note the
```
That link is broken.

levic replied:
How tightly coupled is this to ant?

More specifically, this is the antd widget implementation so I would expect it to be highly dependent on ant, but if I wanted to use another widget lib then could I potentially to see a quite different async choices interface?

davecoates replied:
Reasonably specific to how antd select works. The usage of async choices is specific to ant but the AsyncChoices interface is generic (or at least that's the goal)

levic commented on js-packages/@prestojs/ui-antd/src/widgets/SelectAsyncChoiceWidget.tsx:
```diff
@@ -0,0 +1,468 @@
+const NEXT_PAGE_VALUE = '__nextpage';
+
+/**
+ * A [Select](https://next.ant.design/components/select/) widget that handles [async choices](doc:AsyncChoicesInterface)
```
Comments re link & ant coupling as above

levic commented on django-testbed/frontend/src/models/User.js:
```diff
@@ -24,7 +32,102 @@ function transformAndCacheUser(data) {
+                return (await endpoints.list.execute({ query: { ids: value }, ...deps })).result;
+            },
+        }),
+    }),
```
There seems to be a lot more boilerplate than I would have expected for this -- is this all required or is it testing every option?

Do you have a minimal example for an async choices that would be a lookup into another table? eg a list of Organisations, perhaps filtered by what the user is allowed to see.

davecoates replied:
There's 3 fields defined here to cover different use cases. `referredBy` is the simple use case:

```
referredBy: new IntegerField({
   asyncChoices: new AsyncChoices(asyncChoicesOptions),
 }),
```

in terms of reducing `asyncChoicesOptions` it's fairly generic so I imagine we'll have `ViewModelAsyncChoices` or something like that that does that (may be default for foreign key fields?)

levic commented on js-packages/@prestojs/util/src/Labeled.ts:
```diff
@@ -0,0 +1,110 @@
+/**
+ * Interface for items that provide a plain text label. Implementing this can save you having to pass
+ * explicit functions to label an item in other parts of the system (eg. for [AsyncChoices](doc:AsyncChoices))
```
Am I able to read AsyncChoices documentation or only after it's merged?

davecoates replied:
Normally yes but currently it's failing to build

levic commented on js-packages/@prestojs/util/src/Labeled.ts:
```diff
@@ -0,0 +1,110 @@
+ * @extract-docs
+ * @menu-group Labeled
+ */
+export interface PlainLabeled {
```
Is Plain vs Rich used elsewhere in react?

It makes sense but it's just terminology I haven't seen before. (TextLabelled vs NodeLabelled maybe)?

Beyond that, if a string and a Node are both renderable by React why do we need the distinction?

levic replied:
Edit: is this because some ant components insist on strings instead of renderable nodes?

davecoates replied:
I haven't seen `Rich` or `Plain`

The reason for the distinction is really for when you want to use string interpolation or other text transformation vs just render something

levic replied:
Can we leave it up to the dev then whether it returns a string or a Node?

If they try to do string interpolation on a Node then they'll get an error, but they were the one who chose to return a Node so they can fix it.

davecoates replied:
You wouldn't get an error would just end up with a string like "Welcome [object Object]"

The intention was to provide different guarantees

Rich - suitable to be rendered to screen 
Plain - suitable to be used in an alert box, fill a text box, send via API call etc

You don't think there's advantage in separating them?

levic commented on js-packages/@prestojs/util/src/identifiable.ts:
```diff
@@ -2,50 +2,82 @@ type SingleId = string | number;
+/**
+ * Defines a property `id` to uniquely identify the item
+ */
+interface IdentifiableId {
```
Why does this exist? If we already have `IdentifiablePk` then shouldn't that be the thing used to access the PK, and `id` is just the site-specific implementation of using `id` as the concrete PK?

If another backend was to use `Id` or `MyModel_Id` then wouldn't that require changes to presto?

davecoates replied:
I think it was really because `id` is the intuitive thing if implementing `Identifiable` it's just that it's not suitable for our use case (although would work most of the time). That's where I started from anyway but didn't want to force `ViewModel` to conform to this so added `_pk` as the support. 

I guess we could just change it to `_pk` it just sounds a bit more specific to a database implementation

(this isn't new behaviour in this PR BTW - I just refactored how the type was defined to make it easier to work with internally)

levic commented on js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:
```diff
@@ -0,0 +1,242 @@
+     * (any additional options passed on the `listOptions` prop to `useAsyncChoices`).
+     * Official presto widgets all use `useAsyncChoices` and so will use these parameters.
+     */
+    useListProps(args: any): any;
```
Can you give an example here of what a "dependency" is? How does it differ from the actual data itself?

eg if I have a list of organisations & their countries then is this supposed to query the country list or is that supposed to be a join that is going as part of the data returned by `list()` ?

davecoates replied:
oops I changed the name from `useListDeps` to `useListProps` but still use dependencies in the comment.

The props returned here is anything required by the eg. widget or hooks that use the async choices. So pagination is the common thing:

```ts
useListProps() {
  const paginator = usePaginator(endpoints.list);
  return { paginator };
}
```

It also can be used to transform the props passed in - so if you wanted to change how the keywords from a generic widget are transformed before they are passed through to the async function (eg. the endpoint) you can do this here without having to have a custom widget.

levic commented on js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:
```diff
@@ -0,0 +1,242 @@
+     */
+    useListProps(args: any): any;
+    /**
+     * Function to resolve a list of choices based on the provided params.
```
`params` is where an autocomplete keyword search string would fit in? How would that get injected given that `useListProps()`is passed straight in?

davecoates replied:
See https://github.com/prestojs/prestojs/pull/67/files#diff-a1b1883790df4d0587b34a90efa4a851R120

You can return it from `listProps` otherwise it's passed straight through in `useAsyncChoices`

levic commented on js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:
```diff
@@ -0,0 +1,242 @@
+     * (any additional options passed on the `retrieveOptions` prop to `useAsyncChoices`).
+     * Official presto widgets all use `useAsyncChoices` and so will use these parameters.
+     */
+    useRetrieveProps(args: any): any;
```
Same comment as `useListProps` above

davecoates replied:
Same kind of deal - most obvious usage I had for it was reading from cache always for existing values to avoid unnecessary fetches, eg https://github.com/prestojs/prestojs/pull/67/files#diff-013e735fdb71bd7fe985d673e4eb416cR62

levic commented on js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:
```diff
@@ -0,0 +1,242 @@
+    retrieve(
+        value: ValueType[] | ValueType,
+        deps?: any
+    ): Promise<typeof value extends ValueType[] ? ItemType[] : ItemType>;
```
Same comment as `list` above

levic commented on js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:
```diff
@@ -0,0 +1,242 @@
+     * what is rendered when an item for a value cannot be found (eg. when it's deleted or
+     * when it's loading). The exact details of how this is used depend on the widget.
+     */
+    getMissingLabel(value: ValueType): React.ReactNode;
```
Is this intended to handle `null` values too or would the Widget handle that?

davecoates replied:
No this is just for if a value is present but no label can be found

davecoates replied:
(widget handles it - there's a default on the widget or can be specified as a prop)

davecoates committed:
fix: minor changes from review

davecoates committed:
doc: ViewModelEndpoint improvements

davecoates committed:
fix: Identifiable use _key

Drop support for 'id' and '_pk' and accept only _key. Update
ViewModel to conform to this.

davecoates committed:
fix: Rich => Node, Plain => Text

davecoates committed:
fix: re-activate skipped tests, minor fix

Change to not pass 'value' through to list / getListProps as this invalidates
the execute function every change which triggers unnecssary list calls. No
current use case that requires 'value' so removing it for now.

davecoates committed:
sys: add no-only-test eslint plugin

Avoid committing test.only changes in future

levic commented on js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts:
```diff
@@ -651,6 +657,11 @@ export default function viewModelFactory<T extends FieldsMapping>(
         },
+        _key: {
+            get(): PrimaryKey {
+                return this._pk;
```
Is `_pk` still needed at all? Can we just replace it entirely with `_key`?

davecoates replied:
Mainly because it was already there and used throughout + parallel to django's `record.pk`

i guess it would make sense to remove it and just use _key

levic reviewed (commented):
Looks good other than the question about _key / _pk: is _pk there just because of history or does it have a distinct meaning from _key? (esp does ViewModel really need both?)

levic commented on js-packages/@prestojs/util/src/Labeled.ts:
```diff
@@ -0,0 +1,110 @@
+ * @extract-docs
+ * @menu-group Labeled
+ */
+export function getPlainLabel(item: any): string {
```
That should be `getTextLabel` for consistency with naming above

davecoates committed:
fix: rename getPlainLabel => getTextLabel

davecoates committed:
fix: bad rebase
